File: 2020/Day 4/JavaScript/Solution.js
const fs = require('fs')

function checkID(input) {
    var pid = input.includes('pid:')
    var iyr = input.includes('iyr');
    var eyr = input.includes('eyr');
    var hgt = input.includes('hgt');
    var hcl = input.includes('hcl');
    var ecl = input.includes('ecl');
    var byr = input.includes('byr');

    if (pid && iyr && eyr && hgt && hcl && ecl && byr) {
        return true
    }
    return false
}

function checkIDDetails(newInput) {
    newStringArray = newInput.split(' ')

    pid = false
    iyr = false
    eyr = false
    hgt = false
    hcl = false
    ecl = false
    byr = false

    newStringArray.forEach(item => {
        if (item.includes('pid') && item.length == 13) {
            pid = true
        }
        if (item.includes('iyr') && item.length == 8 && item.split(':')[1] >= 2010 && item.split(':')[1] <= 2020) {
            iyr = true
        }
        if (item.includes('eyr') && item.length == 8 && item.split(':')[1] >= 2020 && item.split(':')[1] <= 2030) {
            eyr = true
        }

        if (item.includes('hgt') && item.includes('in') && item.split(':')[1].split('in')[0] >= 59 && item.split(':')[1].split('in')[0] <= 76) {
            
            hgt = true
        } else if (item.includes('hgt') && item.includes('cm') && item.split(':')[1].split('cm')[0] >= 150 && item.split(':')[1].split('cm')[0] <= 193) {
            
            hgt = true
        }

        let re = new RegExp('#[aA-fF0-9]{6}')
        if (item.includes('hcl') && re.test(item)) {            
            hcl = true
        }
        
        let eyelist = ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth']
        if (item.includes('ecl') && eyelist.includes(item.split(':')[1])) {
            ecl = true
        }
        
        if (item.includes('byr') && item.split(':')[1] >= 1920 && item.split(':')[1] <= 2002 && item.length == 8) {
            
            byr = true
        }    
    })
    
    if (pid && iyr && eyr && hgt && hcl && ecl && byr) {
        return true
    } else { 
        return false
    }

    
}

function PartOne(input) {
    var results = 0
    var newInput = ""

    input.forEach(item => {
        if (item.length === 1) {
            var correct = checkID(newInput)
            if (correct) {
                results++
            }
            newInput = ""
        } else {
            newInput = newInput + " " + item
        }
    })

    return results
}

function PartTwo(input) {
    var results = 0
    var newInput = ""

    input.forEach(item => {
        if (item.length === 1) {
            var correct = checkIDDetails(newInput)
            if (correct) {
                results++
            }
            newInput = ""
        } else {
            newInput = newInput + " " + item
        }
    })

    return results
}

fs.readFile('..\\Input.txt', 'UTF-8', (err, data) => {
    if (err) {
        console.log(err)
        return
    }

    var array = data.split('\n')

    results = PartOne(array)
    results2 = PartTwo(array)

    console.log(`Part One ${results} Part Two ${results2}`);
})

File: 2020/Day 4/Python/Solution.py
import re

with open('..\\input.txt') as file:
    lines = file.readlines()
    
def checkID(passportID, correct = False):
    if 'pid' in passportID and 'iyr' in passportID and 'eyr' in passportID and 'hgt' in passportID and 'hcl' in passportID and 'ecl' in passportID and 'byr' in passportID:
        correct = True
        return correct

    return correct

def checkIDDetails (passportID, correct = False):
    newStringArray = passportID.split(' ')

    pid, iyr, eyr, hgt, byr, hcl, ecl = False, False, False, False, False, False, False

    for item in newStringArray:
        print("Working on " + item)
        if 'pid' in item and len(item) == 13:
            pid = True
        if 'iyr' in item and int(item.split(':')[1]) >= 2010 and int(item.split(':')[1]) <= 2020 and len(item) == 8:
            iyr = True
        if 'eyr' in item and int(item.split(':')[1]) >= 2020 and int(item.split(':')[1]) <= 2030 and len(item) == 8:
            eyr = True
        if 'hgt' in item and 'in' in item and int(re.split("\D", item)[4]) >= 59 and int(re.split("\D", item)[4]) <= 76:
            hgt = True
        elif 'hgt' in item and 'cm' in item and int(re.split("\D", item)[4]) >= 150 and int(re.split("\D", item)[4]) <= 193:
            hgt = True
        if 'hcl' in item and re.search("#[aA-fF0-9]{6}", item):
            hcl = True
        eyes = ("amb", "blu", "brn", "gry", "grn", "hzl", "oth")
        if 'ecl' in item and item.split(':')[1] in eyes:
            ecl = True
        if 'byr' in item and int(item.split(':')[1]) >= 1920 and int(item.split(':')[1]) <= 2002 and len(item) == 8:
            byr = True

        if pid and iyr and eyr and hgt and hcl and ecl and byr:
            
            return True

    return correct


def PartOne():
    results = 0
    newInput = ""

    for ele in lines:
        if ele.isspace() == True:
            correct = checkID(newInput.strip())

            if correct:
                results += 1

            newInput = ""

        else:
            newInput = str(newInput) + " " + str(ele.strip())

    print("Found " + str(results))
        
def PartTwo():
    results = 0
    newInput = ""

    for ele in lines:
        if ele.isspace() == True:
            correct = checkIDDetails(newInput.strip())

            if correct:
                results += 1

            newInput = ""

        else:
            newInput = str(newInput) + " " + str(ele.strip())
    
    print("Found " + str(results))


PartOne()
PartTwo()